File: swift/DesignCircularQueue.swift
// https://leetcode.com/problems/design-circular-queue/


class MyCircularQueue {
    
    private class Node {
        var val: Int
        var next: Node?
        
        init(_ val: Int, _ next: Node? = nil) {
            self.val = val
            self.next = next
        }
    }
    
    private var head: Node? = nil
    private var tail: Node? = nil
    
    private var capacity: Int
    private var size: Int = 0

    init(_ k: Int) {
        capacity = k
    }
    
    func enQueue(_ value: Int) -> Bool {
        guard size < capacity else { return false }
        if tail != nil {
            tail?.next = Node(value)
            tail = tail?.next
        } else {
            head = Node(value)
            tail = head
        }
        size += 1
        return true
    }
    
    func deQueue() -> Bool {
        guard size > 0 else { return false }
        head = head?.next
        size -= 1
        if head == nil {
            tail = nil
        }
        return true
    }
    
    func Front() -> Int {
        return head?.val ?? -1
    }
    
    func Rear() -> Int {
        return tail?.val ?? -1
    }
    
    func isEmpty() -> Bool {
        return size == 0
    }
    
    func isFull() -> Bool {
        return size == capacity
    }
}

File: c/DesignCircularQueue.c
// https://leetcode.com/problems/design-circular-queue/

typedef struct {
    int *arr;
    int capacity, size;
    int head, tail;
} MyCircularQueue;

MyCircularQueue* myCircularQueueCreate(int k) {
    MyCircularQueue* queue = malloc(sizeof(MyCircularQueue));
    queue->arr = (int *)malloc(sizeof(int) * k);
    queue->capacity = k;
    queue->size = 0;
    queue->head = 0;
    queue->tail = -1;
    return queue;
}

bool myCircularQueueEnQueue(MyCircularQueue* obj, int value) {
    if (obj->size < obj->capacity) {
        if (obj->tail + 1 == obj->capacity)
            obj->arr[obj->tail = 0] = value;
        else
            obj->arr[++obj->tail] = value;
        obj->size++;
        return true;
    }
    return false;
}

bool myCircularQueueDeQueue(MyCircularQueue* obj) {
    if (obj->size > 0) {
        if (obj->size == 1) {
            obj->head = 0;
            obj->tail = -1;
        } else if (obj->head + 1 > obj->capacity)
            obj->head = 0;
        else
            obj->head++;
        obj->size--;
        return true;
    }
    return false;
}

int myCircularQueueFront(MyCircularQueue* obj) {
    return obj->size > 0 ? obj->arr[obj->head] : -1;
}

int myCircularQueueRear(MyCircularQueue* obj) {
    return obj->size > 0 ? obj->arr[obj->tail] : -1;
}

bool myCircularQueueIsEmpty(MyCircularQueue* obj) {
    return obj->size == 0;
}

bool myCircularQueueIsFull(MyCircularQueue* obj) {
    return obj->size == obj->capacity;
}

void myCircularQueueFree(MyCircularQueue* obj) {
    free(obj->arr);
    free(obj);
}